### PR TITLE
Add option to run server with a threadpool

### PR DESCRIPTION
**Pull request by jaemk**
issue #167 
- Add crates: `threadpool`, `num_cpus`
- Add `start_server_with_pool` function that mirrors the existing
  `start_server` function
- Add an `Executor` instance to the `Server` to spawn server
  processes in either a `std::thread` or a `threadpool::ThreadPool`
- Change `Server::run` signature to take a reference instead of
  ownership so it mirrors `Server::poll` and can be chained
  after a call to `Server::pool_size`

### CODE REVIEW

**Comment by jaemk** [comment]
Some test results (numbers are req/s) to confirm the default mode isn't affected:

----

```
cargo run --release --example hello-world
wrk -t4 -c100 -d10s http://localhost:8000
```

| current | new - thread | new - threadpool |
|--|--|--|
| 25500 | 25200 | 102200 |

----

```
cargo run --release --example database
# insert some data
curl -X POST -d 'some data' -H 'Content-Type: text/plain' http://locahost:8000/note
wrk -t4 -c100 -d10s http://localhost:8000/notes
```

| current | new - thread | new - threadpool |
|--|--|--|
| 6850 | 6820 | 8100 |

File: src/lib.rs [hunk 1]
```diff
@@ -263,7 +322,7 @@ impl<F> Server<F> where F: Send + Sync + 'static + Fn(&Request) -> Response {
     /// Runs the server forever, or until the listening socket is somehow force-closed by the
     /// operating system.
     #[inline]
-    pub fn run(self) {
+    pub fn run(&self) {
```
Comment (tomaka) [inline src/lib.rs]: What is the reason of this change?

**Comment by tomaka** [comment]
The code looks good. Please also add an entry to CHANGELOG.md.

File: src/lib.rs (hunk 1 above)
Comment (jaemk) [reply src/lib.rs]: This is so a call to `run` can be chained after a `pool_size`:

```rust
Server::new("localhost:8000", handler)?
    .pool_size(8 * num_cpus::get())
    .run()
```

**Comment by jaemk** [comment]
Changelog is updated!

File: src/lib.rs (hunk 1 above)
Comment (tomaka) [reply src/lib.rs]: I'd prefer that `pool_size` takes by `self` and returns a `Self`.

Comment (jaemk) [reply src/lib.rs]: Alright, `pool_size` and `run` signatures have been changed to take/return `self`.
